src/MonitorProfiler/Utilities/NameCache.cpp

```diff
@@ -0,0 +1,223 @@
+{
+    return GetData(_moduleNames, id, data);
+}
+bool NameCache::GetTokenData(ModuleID modId, mdTypeDef token, std::shared_ptr<TokenData>& data)
+{
+    auto const& it = _names.find(std::make_pair(modId, token));
+
+    if (it != _names.end())
+    {
+        data = it->second;
+        return true;
+    }
+    return false;
+}
+
+void NameCache::AddModuleData(ModuleID moduleId, tstring&& name)
+{
+    _moduleNames.emplace(moduleId, std::make_shared<ModuleData>(std::move(name)));
+}
+
+HRESULT NameCache::GetFullyQualifiedName(FunctionID id, tstring& name)
+{
+    HRESULT hr;
+
+    std::shared_ptr<FunctionData> functionData;
+    if (GetFunctionData(id, functionData))
+    {
+        std::shared_ptr<ClassData> classData;
+        if (functionData->GetClass() != 0)
+        {
+            IfFailRet(GetFullyQualifiedClassName(functionData->GetClass(), name));
+        }
+        else
+        {
+            IfFailRet(GetFullyQualifiedClassName(functionData->GetModuleId(), functionData->GetClassToken(), name));
+        }
+
+        name += FunctionSeperator + functionData->GetName();
+
+        for (int i = 0; i < functionData->GetTypeArgs().size(); i++)
+        {
+            if (i == 0)
+            {
+                name += GenericBegin;
+            }
+
+            tstring genericParamName;
+            IfFailRet(GetFullyQualifiedClassName(functionData->GetTypeArgs()[i], genericParamName));
+            name += genericParamName;
+
+            if (i == (functionData->GetTypeArgs().size() - 1))
+            {
+                name += GenericEnd;
+            }
+            else
+            {
+                name += GenericSeparator;
+            }
+        }
+
+        std::shared_ptr<ModuleData> moduleData;
+        if (GetModuleData(functionData->GetModuleId(), moduleData))
+        {
+            name = moduleData->GetName() + ModuleSeparator + name;
+        }
+    }
+
+    return S_OK;
+}
+
+HRESULT NameCache::GetFullyQualifiedClassName(ClassID classId, tstring& name)
+{
+    HRESULT hr;
+
+    std::shared_ptr<ClassData> classData;
+    if (classId != 0 && GetClassData(classId, classData))
+    {
+        mdTypeDef token = classData->GetToken();
+
+        if (classData->GetFlags() == ClassFlags::None)
+        {
+            IfFailRet(GetFullyQualifiedClassName(classData->GetModuleId(), classData->GetToken(), name));
+        }
+        else
+        {
+            switch (classData->GetFlags())
+            {
+                case ClassFlags::Array:
+                    name = ArrayClassName;
+                    break;
+                case ClassFlags::Composite:
+                    name = CompositeClassName;
+                    break;
+                case ClassFlags::IncompleteData:
+                case ClassFlags::Error:
+                default:
+                    name = UnknownName;
+                    break;
+            }
+        }
+
+        for (int i = 0; i < classData->GetTypeArgs().size(); i++)
```

Comment:
It looks like this is a nearly identical copy-paste from the prior method; just from a code cleanliness perspective, is there any way this can be made more generic so we can reuse the `name` construction?